packages/yew-router/src/history.rs

```diff
@@ -0,0 +1,227 @@
+                    self.path = path.into();
+                } else if path.is_empty() {
+                    self.path = "/".into();
+                }
+            }
+        }
+    }
+    pub(crate) fn empty() -> Self {
+        Self {
+            path: Default::default(),
+            query: Default::default(),
+            hash: Default::default(),
+        }
+    }
+}
+
+#[derive(Debug)]
+pub enum HistoryAction {
+    Push(Route),
+    Replace(Route),
+}
+
+struct HistoryState {
+    last_route: Rc<Route>,
+    subscribers: Vec<Callback<Rc<Route>>>,
+    _listeners: [EventListener; 2],
+}
+
+thread_local! {
+    static HISTORY_STATE: RefCell<Option<HistoryState>> = RefCell::new(None);
+}
+
+impl HistoryState {
+    fn with<R>(f: impl FnOnce(&mut Option<Self>) -> R) -> R {
+        HISTORY_STATE.with(|state| f(&mut *state.borrow_mut()))
+    }
+    fn determine_current_route() -> Route {
+        let window = window();
+        let location = window.location();
+        let path = location.pathname().expect("no pathname");
+        let query = location.search().expect("no pathname");
+        let hash = location.hash().expect("no pathname");
+        let mut res = Route { path, query, hash };
+        res.unapply_base();
+        res
+    }
+
+    fn new() -> Self {
+        // Install event listeners
+        let et: EventTarget = window().into();
+        let _listeners = [
+            EventListener::new(&et, "popstate", |_| Self::update()),
+            EventListener::new(&et, "hashchange", |_| Self::update()),
+        ];
+
+        Self {
+            last_route: Rc::new(Self::determine_current_route()),
+            subscribers: Vec::new(),
+            _listeners,
+        }
+    }
+
+    // We sometimes return a function to run when the state is not borrowed.
+    // This is so that callbacks don't panic if they try to access the state.
+    fn update_inner(maybe_state: &mut Option<Self>) -> Option<impl FnOnce() + 'static> {
+        let state = maybe_state.as_mut()?;
+        if state.subscribers.is_empty() {
+            *maybe_state = None;
+            None
+        } else {
+            let route = Rc::new(Self::determine_current_route());
+            if state.last_route != route {
+                state.last_route = route.clone();
+                let subscribers = state.subscribers.clone();
+                Some(move || {
+                    for subscriber in subscribers {
+                        subscriber.emit(route.clone());
+                    }
+                })
+            } else {
+                None
+            }
+        }
+    }
+
+    fn update() {
+        if let Some(f) = Self::with(Self::update_inner) {
+            f();
+        }
+    }
+
+    fn current_route(maybe_state: &mut Option<Self>) -> Rc<Route> {
+        maybe_state.get_or_insert_with(Self::new).last_route.clone()
+    }
+
+    fn register(maybe_state: &mut Option<Self>, callback: Callback<Rc<Route>>) -> HistoryListener {
```

Comment:
`register` isn't very descriptive. Anyone reading the code would wonder, "register what?" I think it should be renamed to `register_listener`

Reply:
👍